gobblin-runtime/src/main/java/gobblin/runtime/mapreduce/MRJobLauncher.java

```diff
@@ -577,17 +581,53 @@ protected void setup(Context context) {
     public void run(Context context)
         throws IOException, InterruptedException {
       this.setup(context);
-
+      GobblinMultiTaskAttempt gobblinMultiTaskAttempt = null;
       try {
         // De-serialize and collect the list of WorkUnits to run
         while (context.nextKeyValue()) {
           this.map(context.getCurrentKey(), context.getCurrentValue(), context);
         }
         // Actually run the list of WorkUnits
-        runWorkUnits(this.jobState.getJobId(), context.getTaskAttemptID().toString(), this.jobState, this.workUnits,
-            this.taskStateTracker, this.taskExecutor, this.taskStateStore, LOG);
+        gobblinMultiTaskAttempt =
+            runWorkUnits(this.jobState.getJobId(), context.getTaskAttemptID().toString(), this.jobState, this.workUnits,
+                this.taskStateTracker, this.taskExecutor, this.taskStateStore, LOG, !this.isSpeculativeEnabled);
+
+        if (this.isSpeculativeEnabled) {
+          LOG.info("will not commit in task attempt");
+          GobblinOutputCommitter gobblinOutputCommitter = (GobblinOutputCommitter) context.getOutputCommitter();
+          gobblinOutputCommitter.getAttemptIdToMultiTaskAttempt()
+              .put(context.getTaskAttemptID().toString(), gobblinMultiTaskAttempt);
+        }
       } finally {
-        this.cleanup(context);
+        CommitStep cleanUpCommitStep = new CommitStep() {
+
+          @Override
+          public boolean isCompleted()
+              throws IOException {
+            throw new UnsupportedOperationException("Not supported.");
```

Comment:
What is the expected behavior here? Isn't it better to return "false" instead of throwing an exception?